crates/ty_python_semantic/src/types/generics.rs

```diff
@@ -128,6 +128,14 @@ impl<'db> GenericContext<'db> {
         Self::new(db, variables)
     }
 
+    /// Creates a generic context from a list of (synthetic) `BoundTypeVarInstance`s.
```

Comment:
nit: I'd suggest removing the "(synthetic)" part, since that doesn't have to be a limitation on when you can use this constructor.  And I think you could simplify `from_type_params` and `normalized_impl` to delegate to this constructor to do the collecting.